### DISCUSSION TITLE
Why doesn't `time::timeout` work with `io::copy`?

### DISCUSSION THREAD

**Discussion by Jimmy-Z** [tokio-rs/tokio#6248] [Q&A]
The following code will run indefinitely instead of stopping after 1s, but why?

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1418b22bb3afc5a36a368f80cbcefc40

the document of `timeout` says:

> if the future does not yield during execution then it is possible for the future to complete and exceed the timeout without returning an error.

I'm not sure what exactly does "yield" mean here, but I'm quite sure `copy` is not blocking.

**Comment by Jimmy-Z** [comment]
A little more test shows if I use `tokio::fs::File::open("/dev/zero")` instead of `repeat`, it will stop, I'm guessing this is because `Repeat::poll_read()`  always returns `Ready`, never `Pending`.

https://docs.rs/tokio/latest/src/tokio/io/util/repeat.rs.html#51-61

I'd appreciate if someone could confirm my guessing though.

**Comment by Darksonn** [comment]
It's because neither `repeat` nor `sink` ever yield, and `copy` has no protection against that case. I've opened #6249 and #6250 to track this issue.

**Comment by Jimmy-Z** [comment]
https://tokio.rs/tokio/glossary#yielding
> a future yields whenever it returns from the poll method.

I think the document need to clarify on the term "yield", clearly that glossary entry doesn't apply here.

**Reply by Darksonn** [reply]
It's accurate for futures, but `repeat` and `sink` are not futures.

(But updating it would be still be reasonable, of course.)

**Reply by Jimmy-Z** [reply]
Well I mean the difference between

> returns from the `poll` method

and 

> returns `Pending` from the `poll` method
